ml-algorithms/src/main/java/org/opensearch/ml/engine/ingest/S3DataIngestion.java

```diff
@@ -0,0 +1,207 @@
+
+        return calculateSuccessRate(successRates);
+    }
+
+    public double ingestSingleSource(
+        S3Client s3,
+        String s3Uri,
+        MLBatchIngestionInput mlBatchIngestionInput,
+        int sourceIndex,
+        boolean isSoleSource
+    ) {
+        String bucketName = getS3BucketName(s3Uri);
+        String keyName = getS3KeyName(s3Uri);
+        GetObjectRequest getObjectRequest = GetObjectRequest.builder().bucket(bucketName).key(keyName).build();
+        double successRate = 0;
+
+        try {
+            ResponseInputStream<GetObjectResponse> s3is = AccessController
+                .doPrivileged((PrivilegedExceptionAction<ResponseInputStream<GetObjectResponse>>) () -> s3.getObject(getObjectRequest));
+            BufferedReader reader = new BufferedReader(new InputStreamReader(s3is, StandardCharsets.UTF_8));
+
+            List<String> linesBuffer = new ArrayList<>();
+            String line;
+            int lineCount = 0;
+            // Atomic counters for tracking success and failure
+            AtomicInteger successfulBatches = new AtomicInteger(0);
+            AtomicInteger failedBatches = new AtomicInteger(0);
+            // List of CompletableFutures to track batch ingestion operations
+            List<CompletableFuture<Void>> futures = new ArrayList<>();
+
+            while ((line = reader.readLine()) != null) {
+                linesBuffer.add(line);
+                lineCount++;
+
+                // Process every 100 lines
+                if (lineCount == 100) {
+                    // Create a CompletableFuture that will be completed by the bulkResponseListener
+                    CompletableFuture<Void> future = new CompletableFuture<>();
+                    batchIngest(
+                        linesBuffer,
+                        mlBatchIngestionInput,
+                        getBulkResponseListener(successfulBatches, failedBatches, future),
+                        sourceIndex,
+                        isSoleSource
+                    );
+
+                    futures.add(future);
+                    linesBuffer.clear();
+                    lineCount = 0;
+                }
+            }
+            // Process any remaining lines in the buffer
+            if (!linesBuffer.isEmpty()) {
+                CompletableFuture<Void> future = new CompletableFuture<>();
+                batchIngest(
+                    linesBuffer,
+                    mlBatchIngestionInput,
+                    getBulkResponseListener(successfulBatches, failedBatches, future),
+                    sourceIndex,
+                    isSoleSource
+                );
+                futures.add(future);
+            }
+
+            reader.close();
+
+            // Combine all futures and wait for completion
+            CompletableFuture<Void> allFutures = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+
+            // Wait for all tasks to complete
+            allFutures.join();
```

Comment:
Which thread will this part block?

Reply:
It's blocking the TRAIN_THREAD_POO. This thread pool was designed to train build-in models offline. But it's almost never used at all, so we can reuse it here for async ingestion.